File: op/openapi/examples/python/openapi_client.py
"""
OpenAPI Client Example (Python)

Requirements:
    pip install requests

Usage:
    client = OpenApiClient(app_id="partner-a", secret="your-secret")
    result = client.get("/fund/selector/list")
    result = client.post("/fund/add", {"name": "Test Fund"})
"""

import base64
import hashlib
import hmac
import time
from typing import Optional, Dict, Any

try:
    import requests
except ImportError:
    raise ImportError("Please install requests: pip install requests")


class OpenApiClient:
    """OpenAPI Client for external system integration"""

    def __init__(
        self,
        gateway_url: str = "http://localhost:8080/openapi",
        app_id: str = "partner-a",
        secret: str = "partner-a-secret-key-change-in-production"
    ):
        """
        Initialize the OpenAPI client

        Args:
            gateway_url: API gateway base URL
            app_id: Application ID assigned by the API provider
            secret: Secret key for HMAC signature
        """
        self.gateway_url = gateway_url.rstrip("/")
        self.app_id = app_id
        self.secret = secret
        self.session = requests.Session()

    def get(
        self,
        path: str,
        params: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Send a GET request

        Args:
            path: Request path (e.g., /fund/selector/list)
            params: Query parameters dictionary

        Returns:
            Response body as string
        """
        return self._request("GET", path, params=params)

    def post(
        self,
        path: str,
        data: Optional[Dict[str, Any]] = None,
        json: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Send a POST request

        Args:
            path: Request path
            data: Form data dictionary
            json: JSON data dictionary

        Returns:
            Response body as string
        """
        return self._request("POST", path, data=data, json=json)

    def _request(
        self,
        method: str,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        data: Optional[Dict[str, Any]] = None,
        json: Optional[Dict[str, Any]] = None
    ) -> str:
        """Execute the HTTP request with HMAC signature"""

        # 1. Generate timestamp (milliseconds)
        timestamp = str(int(time.time() * 1000))

        # 2. Build URL (use path as-is for URL construction)
        url = f"{self.gateway_url}{path}"

        # 3. Build sign path (use full path including gateway prefix for signature)
        # Extract the gateway prefix (e.g., /openapi from http://localhost:8080/openapi)
        from urllib.parse import urlparse
        gateway_prefix = urlparse(self.gateway_url).path  # e.g., /openapi
        sign_path = gateway_prefix + path  # e.g., /openapi/fund/add

        # 3. Build query string (must use URL encoding to match server-side calculation)
        import urllib.parse
        query_string = ""
        if params:
            # Filter None values, keep original order, then URL encode
            filtered_params = [(k, v) for k, v in params.items() if v is not None]
            query_string = urllib.parse.urlencode(filtered_params)

        # 4. Build request body
        body = ""
        body_json = None
        if json is not None:
            import json as json_lib
            body_json = json_lib.dumps(json, ensure_ascii=False)
            body = body_json
        elif data:
            # URL encode form data
            filtered_data = [(k, v) for k, v in data.items() if v is not None]
            body = urllib.parse.urlencode(filtered_data)

        # 5. Calculate signature (use full path including gateway prefix)
        sign_content = f"{timestamp}{method}{sign_path}{query_string}{body}"
        sign = self._hmac_sha256(self.secret, sign_content)

        # 6. Build headers
        headers = {
            "X-App-Id": self.app_id,
            "X-Timestamp": timestamp,
            "X-Sign": sign,
        }

        if json is not None:
            headers["Content-Type"] = "application/json"

        # 7. Send request
        if method == "GET":
            response = self.session.get(url, params=params, headers=headers)
        elif method == "POST":
            if json is not None:
                # Use data= with manually encoded JSON to ensure consistency with signature
                response = self.session.post(url, data=body_json.encode('utf-8'), headers=headers)
            else:
                response = self.session.post(url, data=data, headers=headers)
        else:
            raise ValueError(f"Unsupported HTTP method: {method}")

        return response.text

    @staticmethod
    def _hmac_sha256(secret: str, data: str) -> str:
        """
        Calculate HMAC-SHA256 signature

        Args:
            secret: Secret key
            data: Data to sign

        Returns:
            Base64 encoded signature
        """
        mac = hmac.new(
            secret.encode("utf-8"),
            data.encode("utf-8"),
            hashlib.sha256
        )
        return base64.b64encode(mac.digest()).decode("utf-8")


# ==================== Usage Examples ====================

if __name__ == "__main__":
    client = OpenApiClient()

    # Example 1: GET request
    print("=== GET Request Example ===")
    result = client.get("/fund/selector/list")
    print(result)

    # Example 2: GET request with query parameters
    print("\n=== GET with Parameters Example ===")
    result = client.get("/fund/selector/list", params={"pageSize": "5", "page": 1})
    print(result)

File: op/openapi/examples/csharp/OpenApiClient.cs
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Text;

namespace OpenApiClient;

/// <summary>
/// OpenAPI Client Example (.NET)
/// </summary>
public class OpenApiClient : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly string _gatewayUrl;
    private readonly string _appId;
    private readonly string _secret;

    public OpenApiClient(
        string gatewayUrl = "http://localhost:8080/openapi",
        string appId = "partner-a",
        string secret = "partner-a-secret-key-change-in-production")
    {
        _gatewayUrl = gatewayUrl.TrimEnd('/');
        _appId = appId;
        _secret = secret;
        _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
    }

    public void Dispose()
    {
        _httpClient?.Dispose();
    }

    /// <summary>
    /// Send a GET request
    /// </summary>
    public async Task<string> GetAsync(string path, Dictionary<string, string>? queryParams = null)
    {
        var url = BuildUrl(path, queryParams);
        return await SendAsync("GET", url);
    }

    /// <summary>
    /// Send a POST request with JSON body
    /// </summary>
    public async Task<string> PostAsync<T>(string path, T data) where T : class
    {
        var url = BuildUrl(path);
        return await SendAsync("POST", url, data);
    }

    private string BuildUrl(string path, Dictionary<string, string>? queryParams = null)
    {
        var url = $"{_gatewayUrl}{path}";
        if (queryParams != null && queryParams.Count > 0)
        {
            var queryString = string.Join("&",
                queryParams.OrderBy(kv => kv.Key)
                    .Select(kv => $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value)}"));
            url += "?" + queryString;
        }
        return url;
    }

    private async Task<string> SendAsync<T>(string method, string url, T? data = null) where T : class
    {
        // 1. Generate timestamp (milliseconds since Unix epoch)
        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString();

        // 2. Build request
        using var request = new HttpRequestMessage(new HttpMethod(method), url);
        request.Headers.Add("X-App-Id", _appId);
        request.Headers.Add("X-Timestamp", timestamp);

        if (data != null)
        {
            request.Content = JsonContent.Create(data, options: new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            });

            // 3. Calculate signature (includes body)
            var body = await request.Content.ReadAsStringAsync();
            var signContent = $"{timestamp}{method}{pathFromUrl(url)}{queryFromUrl(url)}{body}";
            request.Headers.Remove("X-Timestamp");
            request.Headers.Add("X-Timestamp", timestamp);
            request.Headers.Add("X-Sign", HmacSha256(_secret, signContent));
        }
        else
        {
            // 3. Calculate signature (no body)
            var signContent = $"{timestamp}{method}{pathFromUrl(url)}{queryFromUrl(url)}";
            request.Headers.Add("X-Sign", HmacSha256(_secret, signContent));
        }

        // 4. Send request
        var response = await _httpClient.SendAsync(request);
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new HttpRequestException($"HTTP {(int)response.StatusCode}: {responseBody}");
        }

        return responseBody;
    }

    private static string pathFromUrl(string url)
    {
        var uri = new Uri(url);
        return uri.AbsolutePath;
    }

    private static string queryFromUrl(string url)
    {
        var uri = new Uri(url);
        return uri.Query.TrimStart('?');
    }

    /// <summary>
    /// Calculate HMAC-SHA256 signature
    /// </summary>
    private static string HmacSha256(string secret, string data)
    {
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
        return Convert.ToBase64String(hash);
    }
}

// ==================== Usage Examples ====================

public class Program
{
    public static async Task Main()
    {
        await using var client = new OpenApiClient();

        // Example 1: GET request
        Console.WriteLine("=== GET Request Example ===");
        var result1 = await client.GetAsync("/fund/selector/list");
        Console.WriteLine(result1);

        // Example 2: GET request with query parameters
        Console.WriteLine("\n=== GET with Parameters Example ===");
        var queryParams = new Dictionary<string, string>
        {
            { "category", "Mixed" },
            { "page", "1" }
        };
        var result2 = await client.GetAsync("/fund/selector/list", queryParams);
        Console.WriteLine(result2);

        // Example 3: POST request
        Console.WriteLine("\n=== POST Request Example ===");
        var requestData = new { name = "Test Fund", category = "Mixed" };
        var result3 = await client.PostAsync("/fund/add", requestData);
        Console.WriteLine(result3);
    }
}
